[Order Details] Add tracking: shipmentProvider shouldn't be saved on dismissal

iamgabrielma opened the pull request:
Partially closes: #1949, together with https://github.com/woocommerce/woocommerce-ios/pull/7156

### Description

The way we currently insert or update the tracking provider within Order Details is not updating properly. The initial intention is to pre-populate the field with the latest selected provider, however, at the moment we find the following behavior:

**Expected:**

- If we select a provider and save it, this provider will appear as prepopulated next time we tap “add tracking”.
- If we select a new provider and save it, this will be updated and appear as prepopulated next time we tap “add tracking”.
- If we select a new provider and dismiss the data, no provider should show if we re-open it.

**What actually happens:**

- Once a provider is saved or dismissed without being saved, we’ll always get back the same provider prepopulating the field, even if we update it and choose a different one.


This PR addresses this problem mainly by changing the logic on upserting data, making our newPreselectedProvider always the first in the list, keeping the file for backward compatibility, as well as moving the call to upsert data into the actual button tapping action, rather than responding to changes as a property observer.

ref: p91TBi-9lY-p2

### Testing instructions
- Install and activate https://woocommerce.com/products/shipment-tracking/
- All the cases can be tested by going to Orders > Select an Order > Scroll to (+) Add tracking :

| Case 1:  Data update | Case 2: Data dismiss |
|---|---|
|![Simulator Screen Shot - iPhone 11 Pro Max - 2022-07-01 at 10 03 56](https://user-images.githubusercontent.com/3812076/176815574-68f8e965-57e2-4391-8c2d-d3180e7e217b.png)  | ![Simulator Screen Shot - iPhone 11 Pro Max - 2022-07-01 at 10 03 39](https://user-images.githubusercontent.com/3812076/176815584-9e02eeac-6437-4715-977f-f84988b52d40.png)|

**Case 1: Data update**
- In your Order, under `Shipping Carrier` add a tracking provider, for example `Fedex`, enter a tracking number, and tap `Add`
- Attempt to add a new provider again by tapping `(+) Add tracking`
- See how `Fedex` is pre-populated.
- Add a new, and different, tracking provider, for example `OnTrac`, enter a tracking number, and tap `Add`
- Attempt to add a new provider again by tapping `(+) Add tracking`
- See how `OnTrac` is pre-populated. 

**Case 2: Data dismiss**
- In your Order, under `Shipping Carrier` add a tracking provider, for example `Fedex`, enter a tracking number, and tap `Dismiss`
- Attempt to add a new provider again by tapping `(+) Add tracking`
- See how nothing is pre-populated

iamgabrielma committed:
Change access control saving and loading selectedShipmentProvider

iamgabrielma committed:
Save selectedShipmentProvider from primaryButtonTapped action

iamgabrielma committed:
Remove saveSelectedShipmentProvider property observer

iamgabrielma committed:
Add additional canCommit check before saving selectedShipmentProvider

iamgabrielma committed:
Change logic on upsert preselectedData

Currently we're not updating correctly the order of elements in dataToSave, with the new logic we attempt to make newPreselectedProvider always the first in the list, as well as keeping the file format for compatibility ref: p91TBi-9lY#comment-10788-p2

iamgabrielma commented on Yosemite/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -488,11 +488,10 @@ private extension AppSettingsStore {
-           let index = preselectedData.firstIndex(of: newPreselectedProvider) {
-            dataToSave[index] = newPreselectedProvider
+        if preselectedData.contains(newPreselectedProvider) {
+            dataToSave[0] = newPreselectedProvider
         } else {
-            dataToSave.append(newPreselectedProvider)
+            dataToSave.insert(newPreselectedProvider, at: 0)
```
Here I opted for updating or inserting depending if the `newPreselectedProvider` can be found in the list or not, however, I believe we don't really need to insert it for what we want to accomplish, we could just always update the new value to `dataToSave[0] = newPreselectedProvider`. What do you think?

joshheald replied:
I agree, but would go one further:

`dataToSave = [newPreselectedProvider]`

Shows that there's only ever one, overwrites all the old data in the file, but doesn't break compatibility.

joshheald commented on WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:
```diff
@@ -237,7 +234,7 @@ final class AddCustomTrackingViewModel: ManualTrackingViewModel {
 
         if returnValue {
-            saveSelectedCustomShipmentProvider()
+            saveSelectedShipmentProvider()
```
I know it's pre-existing, but this still seems like a side-effect, as it was in `didSet`. This is a property which returns a bool on whether we can commit the change.... From the name alone, I would expect to be able to call it repeatedly without that triggering a save.

iamgabrielma committed:
Removed save-tracking side-effects from canCommit

Previously any check for wether the form was ready to be committed would also save the tracking provider, this happen at least twice for each use of the screen and resulted in a tracking save. Removing the save from canCommit means we only save now when the user taps the "add tracking" button.

iamgabrielma committed:
Add guard to viewModel.canCommit

iamgabrielma committed:
Refactor saveTrackingProvider
